### PR TITLE
Unlock Keycard

### PR DESCRIPTION
**Pull request by saledjenic**
Corresponding `StatusQ` PR:
- https://github.com/status-im/StatusQ/pull/904

Corresponding `status-go` PR:
- https://github.com/status-im/status-go/pull/2871

Fixes: #7031 

This one also makes the following issues fully completed:
- #6790  
- #7026 
- #7205

### CODE REVIEW

**Review by stefandunca — COMMENTED** [review]
Testing with an older user folder and trying to reset a locked keycard I end up in a broken state with no other than restarting the app. Maybe broken migration? With a cleanup state, I could reset the keycard though.

Here are screenshot and last log entries, maybe it helps.

image

```log
keycard - card found at index 0
keycard - using reader ACS ACR39U ICC Reader
keycard - card protocol T1
DBG 2022-09-21 22:30:11.087+02:00 keycard_signal                             topics="keycard-service" tid=89789453 file=service.nim:91 response="{\"type\":\"keycard.action.card-inserted\",\"event\":{}}"
DBG 2022-09-21 22:30:11.087+02:00 on_keycard_response                        topics="app-controller" tid=89789453 file=module.nim:273 currFlow=FirstRunOldUserKeycardImport currState=KeycardInsertKeycard
DBG 2022-09-21 22:30:11.091+02:00 on_keycard_response - set state            topics="app-controller" tid=89789453 file=module.nim:278 setCurrFlow=FirstRunOldUserKeycardImport setCurrState=KeycardReadingKeycard
DBG 2022-09-21 22:30:11.349+02:00 keycard_signal                             topics="keycard-service" tid=89789453 file=service.nim:91 response="{\"type\":\"keycard.action.swap-card\",\"event\":{\"error\":\"free-pairing-slots\",\"free-pairing-slots\":0,\"instance-uid\":\"a98fff36348270a0348672901862c24c\",\"key-uid\":\"46d19184a019787fa72f5e8d590a8b9a3caf549debc78ff1262efbf04b41ef29\"}}"
DBG 2022-09-21 22:30:11.350+02:00 on_keycard_response                        topics="app-controller" tid=89789453 file=module.nim:273 currFlow=FirstRunOldUserKeycardImport currState=KeycardReadingKeycard
DBG 2022-09-21 22:30:11.358+02:00 on_keycard_response - set state            topics="app-controller" tid=89789453 file=module.nim:278 setCurrFlow=FirstRunOldUserKeycardImport setCurrState=KeycardMaxPairingSlotsReached
DBG 2022-09-21 22:30:15.303+02:00 primary_action                             topics="app-controller" tid=89789453 file=module.nim:145 currFlow=FirstRunOldUserKeycardImport currState=KeycardMaxPairingSlotsReached
DBG 2022-09-21 22:30:15.307+02:00 service doesn't meant to be used from the context it's used, check the context shared popup module is used topics="keycard-popup-controller" tid=89789453 file=controller.nim:86 service=WalletAccountService
DBG 2022-09-21 22:30:15.307+02:00 sm_there is no any key pair for the logged in user that is not already migrated to a keycard topics="app-controller" tid=89789453 file=module.nim:289
DBG 2022-09-21 22:30:15.308+02:00 keycardCancelFlow                          topics="keycard-service" tid=89789453 file=service.nim:137 currentFlow=-1 response=ok
DBG 2022-09-21 22:30:15.511+02:00 keycardStartFlow                           topics="keycard-service" tid=89789453 file=service.nim:127 currentFlow=13 payload="{\"resolve-addresses\":true}" response=ok
keycard - listing readers
keycard - waiting for card
keycard - card found at index 0
keycard - using reader ACS ACR39U ICC Reader
DBG 2022-09-21 22:30:15.517+02:00 keycard_signal                             topics="keycard-service" tid=89789453 file=service.nim:91 response="{\"type\":\"keycard.action.insert-card\",\"event\":{\"error\":\"connection-error\"}}"
DBG 2022-09-21 22:30:15.519+02:00 shared flow is currently active            topics="app-controller" tid=89789453 file=module.nim:267
keycard - card protocol T1
DBG 2022-09-21 22:30:15.541+02:00 sm_on_keycard_response - from_local - set state topics="app-controller" tid=89789453 file=module.nim:202 setCurrFlow=FactoryReset setCurrState=InsertKeycard
DBG 2022-09-21 22:30:15.543+02:00 keycard_signal                             topics="keycard-service" tid=89789453 file=service.nim:91 response="{\"type\":\"keycard.action.card-inserted\",\"event\":{}}"
DBG 2022-09-21 22:30:15.543+02:00 shared flow is currently active            topics="app-controller" tid=89789453 file=module.nim:267
DBG 2022-09-21 22:30:15.544+02:00 sm_on_keycard_response                     topics="app-controller" tid=89789453 file=module.nim:208 currFlow=FactoryReset currState=InsertKeycard
DBG 2022-09-21 22:30:15.553+02:00 sm_on_keycard_response - set state         topics="app-controller" tid=89789453 file=module.nim:214 setCurrFlow=FactoryReset setCurrState=KeycardInserted
DBG 2022-09-21 22:30:15.761+02:00 keycard_signal                             topics="keycard-service" tid=89789453 file=service.nim:91 response="{\"type\":\"keycard.flow-result\",\"event\":{\"error\":\"unknown-flow\"}}"
DBG 2022-09-21 22:30:15.761+02:00 shared flow is currently active            topics="app-controller" tid=89789453 file=module.nim:267
DBG 2022-09-21 22:30:15.762+02:00 sm_on_keycard_response                     topics="app-controller" tid=89789453 file=module.nim:208 currFlow=FactoryReset currState=KeycardInserted
DBG 2022-09-21 22:30:16.553+02:00 sm_secondary_action                        topics="app-controller" tid=89789453 file=module.nim:168 currFlow=FactoryReset currState=KeycardInserted
DBG 2022-09-21 22:30:16.570+02:00 sm_secondary_action - set state            topics="app-controller" tid=89789453 file=module.nim:175 setCurrFlow=FactoryReset setCurrState=ReadingKeycard
DBG 2022-09-21 22:30:19.555+02:00 sm_secondary_action                        topics="app-controller" tid=89789453 file=module.nim:168 currFlow=FactoryReset currState=ReadingKeycard
```

**Review by stefandunca — COMMENTED** [review]
I suppose [this view](https://www.figma.com/file/idUoxN7OIW2Jpp3PMJ1Rl8/%E2%9A%99%EF%B8%8F-Settings-%7C-Desktop?node-id=199%3A2653) is not finished

image

**Review by stefandunca — COMMENTED** [review]
After having the account set up with a keycard I try to log in with another keycard and used the same. Had the option to enter the pin after which I got stuck in the login animation screen.

Logs that might help


```log
keycard - listing readers
keycard - waiting for card
keycard - card found at index 0
keycard - using reader ACS ACR39U ICC Reader
keycard - card protocol T1
Network Cache Dir:  "/Users/stefan/proj/status/test-data/keycard-test/tmp/netcache"
INF 2022-09-21 22:57:12.695+02:00 starting application...                    topics="status-app" tid=89836190 file=nim_status_client.nim:175
DBG 2022-09-21 22:57:12.857+02:00 keycard_signal                             topics="keycard-service" tid=89836190 file=service.nim:91 response="{\"type\":\"keycard.action.enter-pin\",\"event\":{\"error\":\"\",\"free-pairing-slots\":4,\"instance-uid\":\"4e51cf6526ab7b2bdb9b85a3538cd4c6\",\"key-uid\":\"891e7a5867bf5e25a21cc74e3863284a8bc2e12289d899046b2eb3319f83889f\",\"pin-retries\":3,\"puk-retries\":5}}"
DBG 2022-09-21 22:57:12.857+02:00 on_keycard_response                        topics="app-controller" tid=89836190 file=module.nim:273 currFlow=AppLogin currState=Login
INF 2022-09-21 22:57:13.878+02:00 KeychainManager stopped:                   topics="keychain-service" tid=89836190 file=service.nim:55 errCode=3 errType=authentication errDesc="User did not authenticate successfully: The user tapped the cancel button in the authentication dialog"
DBG 2022-09-21 22:57:13.880+02:00 primary_action                             topics="app-controller" tid=89836190 file=module.nim:145 currFlow=AppLogin currState=Login
DBG 2022-09-21 22:57:13.887+02:00 primary_action - set state                 topics="app-controller" tid=89836190 file=module.nim:151 setCurrFlow=AppLogin setCurrState=LoginKeycardEnterPin
DBG 2022-09-21 22:57:18.952+02:00 tertiary_action                            topics="app-controller" tid=89836190 file=module.nim:171 currFlow=AppLogin currState=LoginKeycardEnterPin
DBG 2022-09-21 22:57:18.952+02:00 keycardCancelFlow                          topics="keycard-service" tid=89836190 file=service.nim:137 currentFlow=-1 response=ok
DBG 2022-09-21 22:57:19.180+02:00 tertiary_action - set state                topics="app-controller" tid=89836190 file=module.nim:177 setCurrFlow=AppLogin setCurrState=WelcomeOldStatusUser
DBG 2022-09-21 22:57:20.166+02:00 secondary_action                           topics="app-controller" tid=89836190 file=module.nim:158 currFlow=AppLogin currState=WelcomeOldStatusUser
DBG 2022-09-21 22:57:20.167+02:00 keycardCancelFlow                          topics="keycard-service" tid=89836190 file=service.nim:137 currentFlow=-1 response="cannot cancel idle flow"
DBG 2022-09-21 22:57:20.369+02:00 keycardStartFlow                           topics="keycard-service" tid=89836190 file=service.nim:127 currentFlow=1 payload="{\"overwrite\":true}" response=ok
keycard - listing readers
keycard - waiting for card
keycard - card found at index 0
keycard - using reader ACS ACR39U ICC Reader
keycard - card protocol T1
DBG 2022-09-21 22:57:21.023+02:00 keycard_signal                             topics="keycard-service" tid=89836190 file=service.nim:91 response="{\"type\":\"keycard.action.enter-pin\",\"event\":{\"error\":\"\",\"free-pairing-slots\":4,\"instance-uid\":\"4e51cf6526ab7b2bdb9b85a3538cd4c6\",\"key-uid\":\"891e7a5867bf5e25a21cc74e3863284a8bc2e12289d899046b2eb3319f83889f\",\"pin-retries\":3,\"puk-retries\":5}}"
DBG 2022-09-21 22:57:21.026+02:00 on_keycard_response                        topics="app-controller" tid=89836190 file=module.nim:273 currFlow=FirstRunOldUserKeycardImport currState=WelcomeOldStatusUser
DBG 2022-09-21 22:57:21.033+02:00 on_keycard_response - set state            topics="app-controller" tid=89836190 file=module.nim:278 setCurrFlow=FirstRunOldUserKeycardImport setCurrState=KeycardEnterPin
DBG 2022-09-21 22:57:23.443+02:00 primary_action                             topics="app-controller" tid=89836190 file=module.nim:145 currFlow=FirstRunOldUserKeycardImport currState=KeycardEnterPin
DBG 2022-09-21 22:57:23.564+02:00 primary_action                             topics="app-controller" tid=89836190 file=module.nim:145 currFlow=FirstRunOldUserKeycardImport currState=KeycardEnterPin
DBG 2022-09-21 22:57:23.629+02:00 primary_action                             topics="app-controller" tid=89836190 file=module.nim:145 currFlow=FirstRunOldUserKeycardImport currState=KeycardEnterPin
DBG 2022-09-21 22:57:23.695+02:00 primary_action                             topics="app-controller" tid=89836190 file=module.nim:145 currFlow=FirstRunOldUserKeycardImport currState=KeycardEnterPin
DBG 2022-09-21 22:57:23.747+02:00 primary_action                             topics="app-controller" tid=89836190 file=module.nim:145 currFlow=FirstRunOldUserKeycardImport currState=KeycardEnterPin
DBG 2022-09-21 22:57:23.802+02:00 primary_action                             topics="app-controller" tid=89836190 file=module.nim:145 currFlow=FirstRunOldUserKeycardImport currState=KeycardEnterPin
DBG 2022-09-21 22:57:23.803+02:00 keycardResumeFlow                          topics="keycard-service" tid=89836190 file=service.nim:132 currentFlow=1 payload="{\"pin\":\"123456\"}" response=ok
DBG 2022-09-21 22:57:26.082+02:00 keycard_signal                             topics="keycard-service" tid=89836190 file=service.nim:91 response="{\"type\":\"keycard.flow-result\",\"event\":{\"eip1581-key\":{\"address\":\"0x165B631C6ED7fD316c60Ac4B6C763E6960369EAF\",\"publicKey\":\"04e23f9f4ceef06feb7e68c351d86bbe1bbacaca8d4866bb2aa7bcead783ed51339e1b015658342dbc51183e3e86735056f54f6446fd28a7a6fa68d3b77587c7de\"},\"encryption-key\":{\"address\":\"0x027b714C44a33D27FCE38224533Be78Ef0135690\",\"publicKey\":\"043b46537d56b24727aadb8dce1d1f756619f05b08b1239ee671a0aad6c514d43e0d5ac2fe813f5c2b408fe164444d109c5874bd5fc1eca9b0dcc120f9f8fb7a2a\",\"privateKey\":\"29b2a1a5bc3d6b50e6457510bd007aa149139e11cc0501799e5849ee7a33c8b8\"},\"key-uid\":\"891e7a5867bf5e25a21cc74e3863284a8bc2e12289d899046b2eb3319f83889f\",\"master-key\":{\"address\":\"0xe74E17D586227691Cb7b64ed78b1b7B14828B034\",\"publicKey\":\"04f13bf63d685e4c0a22ab57e1898b4c0dda7e968451dc50739b42e4211fa7723531ffb19e7da60c369c088861d636b9c24973ecb2568367fbd3386b9cc66d3e2d\"},\"wallet-key\":{\"address\":\"0xe74E17D586227691Cb7b64ed78b1b7B14828B034\",\"publicKey\":\"04f13bf63d685e4c0a22ab57e1898b4c0dda7e968451dc50739b42e4211fa7723531ffb19e7da60c369c088861d636b9c24973ecb2568367fbd3386b9cc66d3e2d\"},\"wallet-root-key\":{\"address\":\"0x419e3052067974EBC68FAafA01cbF0c17C6753E1\",\"publicKey\":\"048ea109c8c9ca67439aff5abf5be1598004282f210e8c2d3b6123ad2e61638bd205b945c7442a1c10b4b84c8a2efc09868ae458785dd62e78643b3a446ed0ed01\"},\"whisper-key\":{\"address\":\"0x1e6ec17F8fa5b836384b21eB277EC16B1149DCD5\",\"publicKey\":\"049b8ccf956e9c29bb80422c33d7826f3576d1c6eb78ea71f2f1c48c239fab3e9cb2b2c3ac7b54043205851485a8e175d0d2b61efa06c814a2c909d620c280410b\",\"privateKey\":\"a48033d0929df823f66482aa5201adf79de5d44165be78a9037fe58735c31ddc\"}}}"
DBG 2022-09-21 22:57:26.086+02:00 on_keycard_response                        topics="app-controller" tid=89836190 file=module.nim:273 currFlow=FirstRunOldUserKeycardImport currState=KeycardEnterPin
DBG 2022-09-21 22:57:26.096+02:00 on_keycard_response - set state            topics="app-controller" tid=89836190 file=module.nim:278 setCurrFlow=FirstRunOldUserKeycardImport setCurrState=Biometrics
DBG 2022-09-21 22:57:27.737+02:00 primary_action                             topics="app-controller" tid=89836190 file=module.nim:145 currFlow=FirstRunOldUserKeycardImport currState=Biometrics
DBG 2022-09-21 22:57:27.759+02:00 Account saved succesfully                  topics="accounts-service" tid=89836190 file=service.nim:163
ERROR[09-21|22:57:27.834] failed to start node and save account    key-uid=0x891e7a5867bf5e25a21cc74e3863284a8bc2e12289d899046b2eb3319f83889f error="UNIQUE constraint failed: settings.synthetic_id"
```

File: src/app/modules/shared_modules/keycard_popup/controller.nim [hunk 1]
```diff
@@ -30,16 +30,21 @@ type
     keychainConnectionIds: seq[UUID]
     connectionKeycardResponse: UUID
     tmpKeycardContainsMetadata: bool
+    tmpCardMetadata: CardMetadata
```
Comment (stefandunca) [inline src/app/modules/shared_modules/keycard_popup/controller.nim:33]: Wondering what is temporary to this data? I see that we set it but when is it discarded?

File: src/app/modules/shared_modules/keycard_popup/internal/wrong_seed_phrase_state.nim [hunk 2]
```diff
@@ -14,18 +14,28 @@ proc delete*(self: WrongSeedPhraseState) =
...
     self.verifiedSeedPhrase = controller.validSeedPhrase(controller.getSeedPhrase()) and
       controller.seedPhraseRefersToSelectedKeyPair(controller.getSeedPhrase())
     if self.verifiedSeedPhrase:
       controller.storeSeedPhraseToKeycard(controller.getSeedPhraseLength(), controller.getSeedPhrase())
     else:
-      controller.setKeycardData(getPredefinedKeycardData(controller.getKeycardData(), PredefinedKeycardData.WrongSeedPhrase, add = true))
+      controller.setKeycardData(updatePredefinedKeycardData(controller.getKeycardData(), PredefinedKeycardData.WrongSeedPhrase, add = true))
+  if self.flowType == FlowType.UnlockKeycard:
+    controller.setKeycardData(updatePredefinedKeycardData(controller.getKeycardData(), PredefinedKeycardData.WrongSeedPhrase, add = false))
+    sleep(500) # just to shortly remove text on the UI side
```
Comment (stefandunca) [inline src/app/modules/shared_modules/keycard_popup/internal/wrong_seed_phrase_state.nim:27]: Why is this required? Why do we need to shortly remove text from the UI when executing the primary command?

**Review by stefandunca — APPROVED** [review]
I did a line-by-line review and it looks good. I miss too much context to look thoroughly for logic errors.

Where can I find what is keycard data? Is this coming from the library? What is Predefined key card data and why is predefined? Do we keep the `PredefinedKeycardData` on

**Comment by saledjenic** [comment]
> Testing with an older user folder and trying to reset a locked keycard I end up in a broken state with no other than restarting the app. Maybe broken migration? With a cleanup state, I could reset the keycard though.
Here are screenshot and last log entries, maybe it helps.

@stefandunca thanks for this, this response `{\"type\":\"keycard.flow-result\",\"event\":{\"error\":\"unknown-flow\"}}` is something I see for the first time. `unknown-flow` was not handled at all, will ask Michele about that since he developed the keycard lib.

**Comment by saledjenic** [comment]
> I suppose [this view](https://www.figma.com/file/idUoxN7OIW2Jpp3PMJ1Rl8/%E2%9A%99%EF%B8%8F-Settings-%7C-Desktop?node-id=199%3A2653) is not finished

@stefandunca actually it is :) that view is [this one in figma](https://www.figma.com/file/17fc13UBFvInrLgNUKJJg5/Kuba%E2%8E%9CDesktop?node-id=13653%3A451025)

File: src/app/modules/shared_modules/keycard_popup/controller.nim (hunk 1 above)
Comment (saledjenic) [reply src/app/modules/shared_modules/keycard_popup/controller.nim:33]: @stefandunca it's set and discarded when the flow is done, cause it's part of the shared module which is created in the moment when we need to run some flow and destroyed after the flow is done.

File: src/app/modules/shared_modules/keycard_popup/internal/wrong_seed_phrase_state.nim (hunk 2 above)
Comment (saledjenic) [reply src/app/modules/shared_modules/keycard_popup/internal/wrong_seed_phrase_state.nim:27]: @stefandunca cause we need to give a user feeling that something is happening in case the previously entered seed phrase was wrong and he enters new one and click on next button, we need to remove warning and either proceed or display a warning message again.

**Review by osmaczko — APPROVED** [review]
I just briefly look at each file, without a deeper understanding. One alarming thing I noticed is `sleep` calls in nim code, using `sleep` is an antipattern.

**Review by glitchminer — APPROVED** [review]
Approved with caveat that further testing required after merge.
